File: src/backend/modules/m5_production/schemas/work_orders.py
"""
M5 생산/SCM — 작업지시서 Pydantic 스키마
"""
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, Field


class WorkOrderCreate(BaseModel):
    """작업지시서 생성 (계획생산)"""
    order_type: str = Field("make_to_stock", description="유형 (make_to_order/make_to_stock)")
    product_id: str = Field(..., description="생산 품목 ID")
    bom_id: Optional[str] = Field(None, description="BOM ID (없으면 자동 선택)")
    planned_qty: float = Field(..., gt=0, description="계획 수량")
    start_date: Optional[date] = None
    due_date: date = Field(..., description="납기일")
    assigned_to: Optional[str] = Field(None, description="담당자 ID")
    notes: Optional[str] = None


class WorkOrderUpdate(BaseModel):
    """작업지시서 수정 (pending 상태만)"""
    planned_qty: Optional[float] = Field(None, gt=0)
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    assigned_to: Optional[str] = None
    bom_id: Optional[str] = None
    notes: Optional[str] = None


class WorkOrderStatusUpdate(BaseModel):
    """작업지시서 상태 변경"""
    status: str = Field(..., description="변경할 상태 (in_progress/qc_wait/completed)")


class WorkOrderProgressUpdate(BaseModel):
    """생산 수량 보고"""
    produced_qty: float = Field(..., ge=0, description="생산 완료 수량")


class WorkOrderListItem(BaseModel):
    """작업지시서 목록 항목"""
    id: str
    wo_no: str
    order_type: str
    product_name: Optional[str] = None
    product_code: Optional[str] = None
    planned_qty: float
    produced_qty: float = 0
    progress_pct: int = 0
    status: str
    due_date: date
    assigned_to_name: Optional[str] = None
    order_no: Optional[str] = None
    created_at: Optional[datetime] = None


class WorkOrderResponse(BaseModel):
    """작업지시서 상세"""
    id: str
    wo_no: str
    order_type: str
    product_id: str
    product_name: Optional[str] = None
    product_code: Optional[str] = None
    bom_id: Optional[str] = None
    planned_qty: float
    produced_qty: float = 0
    progress_pct: int = 0
    status: str
    start_date: Optional[date] = None
    due_date: date
    assigned_to: Optional[str] = None
    assigned_to_name: Optional[str] = None
    order_id: Optional[str] = None
    order_no: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None

File: src/frontend/src/api/production/workOrders.ts
/**
 * M5 생산/SCM — 작업지시서 API 호출 함수
 */
import api from '../client';

/* ── 타입 정의 ── */

export interface WorkOrder {
  id: string;
  wo_no: string;
  order_type: string;
  product_id: string;
  product_name?: string;
  product_code?: string;
  bom_id?: string;
  planned_qty: number;
  produced_qty: number;
  progress_pct: number;
  status: string;
  start_date?: string;
  due_date: string;
  assigned_to?: string;
  assigned_to_name?: string;
  order_id?: string;
  order_no?: string;
  notes?: string;
  created_at?: string;
}

export interface WorkOrderFormData {
  order_type?: string;
  product_id: string;
  bom_id?: string;
  planned_qty: number;
  start_date?: string;
  due_date: string;
  assigned_to?: string;
  notes?: string;
}

export interface FromOrderResult {
  order_no: string;
  work_orders: { wo_no: string; product_name: string; qty: number }[];
  material_shortage: {
    product_id: string;
    product_name: string;
    product_code: string;
    required_qty: number;
    current_qty: number;
    shortage_qty: number;
  }[];
  has_shortage: boolean;
}

/* ── API 호출 ── */

export async function listWorkOrders(params?: {
  status?: string;
  order_type?: string;
  search?: string;
  page?: number;
  size?: number;
}) {
  const res = await api.get('/production/work-orders', { params });
  return res.data.data;
}

export async function getWorkOrder(id: string) {
  const res = await api.get(`/production/work-orders/${id}`);
  return res.data.data;
}

export async function createWorkOrder(data: WorkOrderFormData) {
  const res = await api.post('/production/work-orders', data);
  return res.data;
}

export async function updateWorkOrder(id: string, data: Partial<WorkOrderFormData>) {
  const res = await api.put(`/production/work-orders/${id}`, data);
  return res.data;
}

export async function createFromOrder(orderId: string) {
  const res = await api.post(`/production/work-orders/from-order/${orderId}`);
  return res.data.data as FromOrderResult;
}

export async function updateWorkOrderStatus(id: string, newStatus: string) {
  const res = await api.patch(`/production/work-orders/${id}/status`, null, {
    params: { new_status: newStatus },
  });
  return res.data;
}

export async function updateWorkOrderProgress(id: string, producedQty: number) {
  const res = await api.patch(`/production/work-orders/${id}/progress`, {
    produced_qty: producedQty,
  });
  return res.data;
}
